src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs

```diff
@@ -30,6 +30,17 @@ public WindowsServiceLifetime(IHostEnvironment environment, IHostApplicationLife
             ThrowHelper.ThrowIfNull(optionsAccessor);
             ThrowHelper.ThrowIfNull(windowsServiceOptionsAccessor);
 
+            // REVIEW: Should we just log a warning instead?
+            // Only validate content root for the new IServiceCollection overload of UseWindowsServices() for maximum backwards compatibility.
+            // The IHostBuilder overload sets the content root unlike the IServiceCollection overload, so it should be less likely to be wrong anyway.
+            // We could try normalizing the paths before comparing, but if the ContentRootPath has changed from it's AppContext.BaseDirectory default,
+            // there's probably a misconfiguration.
+            if (windowsServiceOptionsAccessor.Value.ValidateContentRoot && !string.Equals(environment.ContentRootPath, AppContext.BaseDirectory, StringComparison.Ordinal))
```

Comment:
This is indeed a problem. The CWD is the default, and this is the C:\Windows\system32 folder for Windows services. `AppContext.BaseDirectory` is the `DisableDefaults` behavior and is what we want. C:\Windows\system32 is not a reasonable content root, so we have to do something.

We can change the exception message reference `HostApplicationBuilderSettings.DisableDefaults`, but we cannot be sure if this is extension method is being used with a `HostApplicationBuilder`. The other problem is that it's not always the easiest to observe exceptions thrown from Windows Services.

I think we might have to add a `CreateWindowsServiceApplicationBuilder(WindowsServiceSettings)` method that returns a `HostApplicationBuilder` preconfigured with the right content root, lifetime and loggers. I'm not sure if we would want to follow a similar pattern for systemd even though we don't need to change the content root just to be consistent.

@Tratcher Do you have any ideas? The fundamental problem is that we cannot change the content root of `HostApplicationBuilder` after it's constructed in `UseWindowsService()`.

More context:

- https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-6.0&tabs=visual-studio#current-directory-and-content-root
- https://github.com/dotnet/extensions/pull/1083#issuecomment-462994671

Reply:
Autodetect when running in a service then change the default. That's the most user friendly thing to do.

Reply:
> The fundamental problem is that we cannot change the content root of HostApplicationBuilder after it's constructed

Maybe we should examine if this limitation is strictly necessary. What would be the harm if this was mutable? Yes someone could have read it before it was modified, but that's how every normal .NET object works.

Reply:
> I think we might have to add a `CreateWindowsServiceApplicationBuilder(WindowsServiceSettings)` method that returns a `HostApplicationBuilder` preconfigured with the right content root, lifetime and loggers. I'm not sure if we would want to follow a similar pattern for systemd even though we don't need to change the content root just to be consistent.

That doesn't seem to combine well. UseWindowsService was supposed to light up when running in that environment, having to create a separate builder defeats that.

Can we put an extension method UseWindowsServiceSettings on HostApplicationBuilderSettings and have it conditionally set the ContentRootPath? Then you can recommend that in this error message.

Reply:
> > The fundamental problem is that we cannot change the content root of HostApplicationBuilder after it's constructed
> 
> Maybe we should examine if this limitation is strictly necessary. What would be the harm if this was mutable? Yes someone could have read it before it was modified, but that's how every normal .NET object works.

By the time the `HostApplicationBuilder` is constructed, we've already read `appsettings.json` and bootstrapped our environment. We'd have to update `JsonConfigurationSource`/`FileConfigurationSource` to stop caching the `IFileProvider` and/or, make `PhysicalFileProvider` rerootable. Handling the `IChangeToken`'s might be tricky with a rerootable `PhysicalFileProvider`, but it's probably possible. We'd also have to make sure `HostApplicationBuilder.Environment` gets updated.

The other question is what would the API for updating the content root be? It can't be modifying `HostApplicationBuilder.Configuration[HostDefaults.ContentRootKey]` because that would create a circular dependency. We could consider reacting when apps invoke the `Environment.EnvironmentName/ApplicationName/ContentRootPath/ContentRootFileProvider` setters, but allowing the `IFileProvider` itself to be updated means we would have to stop caching at the `FileConfigurationSource` level not just the `PhysicalFileProvider` level.

So far with `WebApplicationBuilder`, we've prevented the modification of the `WebRootPath` in addition to the `ContentRootPath` and the `WebRootPath` is usually dependent on the `ContentRootPath`. We'd have to think about how to rebase the `WebRootPath` when the `ContentRootPath` updates, but only if it was kept default or set to a relative path.  We'd also need to rerun the default `UseStaticWebAssets()` logic. This way leads to madness. See https://github.com/dotnet/aspnetcore/pull/37033 and eventually https://github.com/dotnet/aspnetcore/pull/37147 for prior attempts a dealing with this.

Another smaller thing to consider is `IHostingStartup`s. Today, the run as part of the `WebApplicationBuilder` constructor. This is important because it can do things like add configuration sources visible to `WebApplicationBuilder.Configuration`. These  `IHostingStartup` are guaranteed  to read the environment before the app has a chance to modify it unless the developer really goes out of the way to disable running these automatically and run them manually.

I really would like to remove this limitation because it is really unfortunate for scenarios such as this, but I'm not sure it's worth all the complexity it would bring to the implementation and how hard it would make it to reason about the behavior.

> That doesn't seem to combine well. UseWindowsService was supposed to light up when running in that environment, having to create a separate builder defeats that.

Agreed.

> Can we put an extension method UseWindowsServiceSettings on HostApplicationBuilderSettings and have it conditionally set the ContentRootPath? Then you can recommend that in this error message.

We could. This still isn't great because it's not the easiest to observe exceptions thrown from Windows Services before the `EventLogLoggerProvider` is configured. It might also be confusing if called on an `IServiceCollection` belonging to an `HostBuilder` instead of an `HostApplicationBuilder`.

> Autodetect when running in a service then change the default. That's the most user friendly thing to do.

This is my preference given the option. This would mean type forwarding almost the entirety of `Microsoft.Extensions.Hosting.WindowsServices` and `System.ServiceProcess.ServiceController` to `Microsoft.Extensions.Hosting`. Are we okay with that?

Reply:
> Autodetect when running in a service then change the default. That's the most user friendly thing to do.

No I mean specifically for the content root, when running on windows, using systems APIs to detect (so no dependency) and changing the content root to be the app base instead.

Basically, if the CWD is the "System folder" then change it to be the app base.

Reply:
Windows services run in a shared process and CWD is `C:\WINDOWS\system32` or SystemDrive\Windows\System32. This is no different from what we do today to [turn on the event log for errors by default on windows](https://github.com/dotnet/runtime/blob/8823eefe91f6b774961c32800dea977b687d7328/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L253-L257)

Reply:
Could we use `WindowsServiceHelpers.IsWindowsService()` and by extension `Win32.GetParentProcess()` to determine whether or not to use the CWD or `AppContext.BaseDirectory`? Or do you think it's really better to compare the CWD against the system32 directory? We could just share the source.

Reply:
And would we make the same change to `HostBuilder`?

Reply:
> Could we use WindowsServiceHelpers.IsWindowsService() and by extension Win32.GetParentProcess() to determine whether or not to use the CWD or AppContext.BaseDirectory? Or do you think it's really better to compare the CWD against the system32 directory? We could just share the source.

I was thinking about the most robust way to do this. Maybe checking for windows services is cleaner but less general purpose. I think we should do this in the "default" host builder scenario.

Reply:
> After normalizing the path of course (PS: I wish we had that built in with strongly typed Path objects).

I really wish we had a reliable way to compare path equality. I get this is a hard problem with various filesystems affecting how paths are normalized. I imagine hard and soft links don't make things any easier particularly if the path points to a file or directory that hasn't been created yet. That's why it would be nice to have an API instead of trying to do this ourselves like we did in `WebApplicationBuilder`.

Reply:
Agree. I think for now you'll have to do something similar to what's in `WebApplicationBuilder`

Reply:
There's no fool proof path comparison, you can even have a filesystem or hardware hide this information, you can run into opaque links (imagine multiple machines and network shares, mapped drives, etc), or permissions which prevent tracing all the paths.  Best to approximate with normalization then comparison.  Some platforms/filesystems could permit a "better" comparison (usually by opening the file and asking the filesystem for more information about it) - but you'll never get it 100% and the cost of getting higher than an approximation is usually too high for the value you get and has side effects (sharing / access denied / etc).  cc @adamsitnik @JeremyKuhne

Reply:
You can open the issue here:

https://github.com/dotnet/docs/issues/new?template=breaking-change.yml

That will then get converted into documentation.

Reply:
I ended up writing it up here: https://github.com/dotnet/docs/issues/29945. Please review and edit as necessary.